BoostTestUi/DevEnv.cpp

```diff
@@ -7,35 +7,86 @@
 
 #include "stdafx.h"
 #include "DevEnv.h"
+#include <vector>
 
-namespace {
+#pragma warning(disable: 4278)
+#pragma warning(disable: 4146)
+//The following #import imports EnvDTE based on its LIBID.
+#import "libid:80cc9f66-e7d8-4ddd-85b6-d9e6cd0e93e2" version("8.0") lcid("0") raw_interfaces_only named_guids
+#pragma warning(default: 4146)
+#pragma warning(default: 4278)
 
-template <typename Itf>
-CComPtr<Itf> GetActiveObject(const CLSID& clsid)
-{
-	CComPtr<IUnknown> pIUnk;
-	HRESULT hr = ::GetActiveObject(clsid, nullptr, &pIUnk);
-	CComPtr<Itf> pItf;
-	if (SUCCEEDED(hr))
-		hr = pIUnk->QueryInterface(&pItf);
-	return pItf;
-}
+typedef CComPtr<EnvDTE::_DTE> DTEPtr;
+typedef std::pair<std::wstring, DTEPtr> DTEInfo;
+typedef std::vector<DTEInfo> DTEInfos;
 
-CLSID ClsidFromProgId(const std::wstring& progId)
+namespace {
+
+DTEInfos getDTEInfos()
 {
-	CLSID clsid = CLSID_NULL;
-	CLSIDFromProgID(progId.c_str(), &clsid);
-	return clsid;
+	DTEInfos dteInfos;
+	CComPtr<IRunningObjectTable> pRot;
+	CComPtr<IEnumMoniker> pEnumMoniker;
+	CComPtr<IMoniker> pMoniker;
+
+	HRESULT hr = GetRunningObjectTable(0, &pRot);
+	if (SUCCEEDED(hr) && pRot)
+		hr = pRot->EnumRunning(&pEnumMoniker);
+
+	if (SUCCEEDED(hr) && pEnumMoniker)
+		while (pEnumMoniker->Next(1, &pMoniker, nullptr) == S_OK)
+		{
+			CComPtr<IUnknown> pObj;
+			pRot->GetObject(pMoniker, &pObj);
+			pMoniker.Release();
+
+			DTEPtr pDte = CComQIPtr<EnvDTE::_DTE>(pObj);
```

Comment:
If a Visual Studio is running and does still implement the rather old interface, this should yield a valid interface pointer for one of the objects in the running object table.